### PR TITLE
changes snapshots to zip database from node datadir

### PR DESCRIPTION
**Pull request by hughy**
## Summary

- updates service:snapshot to zip the entire 'databases/<database>' directory
  and upload to s3 for the configured database
- updates chain:import to overwrite database in datadir with downloaded
  snapshot
- fixes downloads to ensure that file is downloaded and hashed completely
- changes both commands to use local flags instead of remote since this method
  requires the node to be local
- updates unit tests for changes to logic

both taking a snapshot and importing require the node to be stopped.

## Testing Plan

- ran `service:snapshot` on databases big and small
- ran `chain:import` to sync new node datases
- updates to unit tests

## Breaking Change

Is this a breaking change? If yes, add notes below on why this is breaking and
what additional work is required, if any.

```
[ ] Yes
```

### CODE REVIEW

File: ironfish-cli/src/commands/chain/import.ts [hunk 1]
```diff
@@ -55,20 +57,15 @@ export default class ImportSnapshot extends IronfishCommand {
...
       if (!flags.confirm) {
         this.log(
-          `This snapshot (${
-            manifest.file_name
-          }) contains the Iron Fish blockchain up to block ${
-            manifest.block_height
-          }. The size of the latest snapshot file is ${FileUtils.formatFileSize(
-            manifest.file_size,
-          )}`,
+          `This snapshot (${manifest.file_name}) contains the Iron Fish blockchain up to block ${manifest.block_height}. The size of the latest snapshot file is ${fileSize}`,
```
Comment (NullSoldier) [inline ironfish-cli/src/commands/chain/import.ts:68]: I know it was this way before, so this is just a note for more work, we should rename height -> sequence in the manifest and the logs. We call it sequence in our code base and our logs. We can do that in another PR.

File: ironfish-cli/src/commands/chain/import.ts [hunk 2]
```diff
@@ -123,17 +122,17 @@ export default class ImportSnapshot extends IronfishCommand {
...
+    const chainDatabasePath = this.sdk.fileSystem.resolve(this.sdk.config.chainDatabasePath)
 
-    for (const file of files) {
-      const blocks = await fsAsync.readFile(path.join(blockExportPath, file))
-      await client.importSnapshot({ blocks })
-    }
+    CliUx.ux.action.start(
+      `Copying snapshot from ${snapshotDatabasePath} to ${chainDatabasePath}`,
+    )
+    await fsAsync.cp(snapshotDatabasePath, chainDatabasePath, { recursive: true, force: true })
```
Comment (NullSoldier) [inline ironfish-cli/src/commands/chain/import.ts:134]: Should we do a move instead of a copy? That way only the disk inodes need to change and not to copy over 30 GB or more of files?

File: ironfish-cli/src/commands/chain/import.ts [hunk 3]
```diff
@@ -87,31 +84,33 @@ export default class ImportSnapshot extends IronfishCommand {
...
+      // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-unsafe-call
+      response.data.on('data', (chunk: { length: number }) => {
+        downloaded += chunk.length
+        bar.update(downloaded, { downloadedSize: FileUtils.formatFileSize(downloaded) })
+      })
       // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-unsafe-call
       response.data.pipe(writer)
-      hasher.update(response.data)
+      // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-unsafe-call
+      response.data.pipe(hasher)
```
Comment (NullSoldier) [inline ironfish-cli/src/commands/chain/import.ts:112]: Does this work to pipe this this way, or is the response stream consumed by the first pipe?

File: ironfish-cli/src/commands/chain/import.ts [hunk 4]
```diff
@@ -87,31 +84,33 @@ export default class ImportSnapshot extends IronfishCommand {
...
-          const percentage = Math.floor((progressEvent.loaded / progressEvent.total) * 100)
-          bar.update(percentage)
-        },
       })
 
+      // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-unsafe-call
+      response.data.on('data', (chunk: { length: number }) => {
+        downloaded += chunk.length
+        bar.update(downloaded, { downloadedSize: FileUtils.formatFileSize(downloaded) })
+      })
```
Comment (NullSoldier) [inline ironfish-cli/src/commands/chain/import.ts:108]: We need to listen to error too and handle that from the request, especially since this is a large file.

File: ironfish-cli/src/commands/service/snapshot.ts [hunk 5]
```diff
@@ -108,48 +93,23 @@ export default class CreateSnapshot extends IronfishCommand {
...
-      if (result.buffer && result.seq) {
-        const blockFilePath = this.sdk.fileSystem.join(blockExportPath, `${result.seq}`)
-        await fsAsync.writeFile(blockFilePath, Buffer.from(result.buffer))
-        progress.update(result.seq || 0)
-      }
-    }
-
-    progress.stop()
+    const chainInfoResponse = await client.getChainInfo()
+    const blockHeight = Number(chainInfoResponse.content.currentBlockIdentifier.index)
```
Comment (NullSoldier) [inline ironfish-cli/src/commands/service/snapshot.ts:103]: This shouldn't need to be casted to a number.. look into this

File: ironfish-cli/src/commands/service/snapshot.ts [hunk 6]
```diff
@@ -108,48 +93,23 @@ export default class CreateSnapshot extends IronfishCommand {
...
 
     const timestamp = Date.now()
 
     const snapshotFileName = `ironfish_snapshot_${timestamp}.tar.gz`
```
Comment (NullSoldier) [inline ironfish-cli/src/commands/service/snapshot.ts:107]: Do we need to have multiple snapshots here? What if we have just 1 path and delete the one that exists if it already exists. You'll need to clean up all the snapshots otherwise but having 1 will reduce the need for that.

File: ironfish-cli/src/commands/service/snapshot.ts [hunk 7]
```diff
@@ -108,48 +93,23 @@ export default class CreateSnapshot extends IronfishCommand {
...
     const snapshotPath = this.sdk.fileSystem.join(exportDir, snapshotFileName)
 
-    this.log(`Zipping\n    SRC ${blockExportPath}\n    DST ${snapshotPath}\n`)
-    CliUx.ux.action.start(`Zipping ${blockExportPath}`)
-    await this.zipDir(blockExportPath, snapshotPath)
+    this.log(`Zipping\n    SRC ${chainDatabasePath}\n    DST ${snapshotPath}\n`)
+    CliUx.ux.action.start(`Zipping ${chainDatabasePath}`)
+    await this.zipDir(chainDatabasePath, snapshotPath, [databaseLockPath])
```
Comment (NullSoldier) [inline ironfish-cli/src/commands/service/snapshot.ts:112]: It's not safe to exclude the lock file. That should be unnecessary. For this we'll need to open the DB to ensure that no one else has opened it. The reason is, that it's not safe to zip a leveldb database if writes are inflight. It's almost guaranteed to become corrupted if that's the case. You can use waitForOpen() and then close the node afterwards to guarantee we have exclusive access to the node databases.

File: ironfish-cli/src/commands/chain/import.ts (hunk 3 above)
Comment (hughy) [reply ironfish-cli/src/commands/chain/import.ts:112]: It works this way! I also wondered about this, but checksums match on import

File: ironfish-cli/src/commands/chain/import.ts (hunk 2 above)
Comment (hughy) [reply ironfish-cli/src/commands/chain/import.ts:134]: Yeah a move would be better -- faster to rename and doesn't leave an extra 30GB sitting around

File: ironfish-cli/src/commands/chain/import.ts (hunk 1 above)
Comment (hughy) [reply ironfish-cli/src/commands/chain/import.ts]: Updated to sequence

File: ironfish-cli/src/commands/service/snapshot.ts (hunk 6 above)
Comment (hughy) [reply ironfish-cli/src/commands/service/snapshot.ts]: Changed to always use the same default filename with an option to set a different one

File: ironfish-cli/src/commands/chain/import.ts [hunk 8]
```diff
@@ -55,20 +63,15 @@ export default class ImportSnapshot extends IronfishCommand {
         this.exit(1)
```
Comment (NullSoldier) [inline ironfish-cli/src/commands/chain/import.ts:63]: ^ up here were writing to a new tmp directory each attempt

File: ironfish-cli/src/commands/service/snapshot.ts [hunk 9]
```diff
@@ -71,20 +76,13 @@ export default class CreateSnapshot extends IronfishCommand {
...
-        : Number(process.env.MAX_BLOCKS_PER_SNAPSHOT_CHUNK),
-      description: 'The max number of blocks per file in the zipped snapshot',
-    }),
   }
 
   async start(): Promise<void> {
     const { flags } = await this.parse(CreateSnapshot)
 
     const bucketUrl = (flags.bucketUrl || DEFAULT_SNAPSHOT_BUCKET_URL || '').trim()
+    const snapshotFileName = (flags.snapshotFileName || DEFAULT_SNAPSHOT_FILE_NAME).trim()
```
Comment (NullSoldier) [inline ironfish-cli/src/commands/service/snapshot.ts:85]: I think we can get rid of this config, and just use a name here so on disk it has a static name, but in the bucket the snapshot should probably use the timestamp like it was originally.

**Review by NullSoldier — APPROVED** [review]

**Comment by NullSoldier** [comment]
We should write the database version in the manifest and check that after downloading the manifest so we don't download a database that is newer than our code base. You can check that version here, if were less than this https://github.com/iron-fish/ironfish/blob/master/ironfish/src/blockchain/blockchain.ts#L59

File: ironfish-cli/src/commands/chain/import.ts [hunk 10]
```diff
@@ -87,31 +90,39 @@ export default class ImportSnapshot extends IronfishCommand {
       const bar = CliUx.ux.progress({
         barCompleteChar: '\u2588',
         barIncompleteChar: '\u2591',
-        format: 'Downloading snapshot: [{bar}] {value}% | ETA: {eta}s',
+        format:
+          'Downloading snapshot: [{bar}] {percentage}% | {downloadedSize} / {fileSize} | ETA: {eta}s',
```
Comment (NullSoldier) [inline ironfish-cli/src/commands/chain/import.ts:94]: We can use a meter like rescan, and get a better estimate and render it using renderEstimate()

File: ironfish/src/blockchain/blockchain.ts [hunk 11]
```diff
@@ -139,6 +139,10 @@ export class Blockchain {
     this._genesis = newGenesis
   }
 
+  get databaseVersion(): number {
+    return DATABASE_VERSION
```
Comment (NullSoldier) [inline ironfish/src/blockchain/blockchain.ts:143]: So this isn't exactly the DB version, which is important to point out. We don't need this setter because you can use `this.db.getVersion()`. So just to be clear, this is the software DB version, but the `db.getVersion()` is the real DB version that it's migrated to. The difference is, if you haven't started your node and run migrations then this will report a different DB version than the actual DB version of the database.

File: ironfish/src/rpc/routes/chain/importChain.ts [hunk 12]
```diff
@@ -74,7 +75,7 @@ function deserializeChunk(node: IronfishNode, reader: bufio.BufferReader): Block
   for (let i = 0; i < blocksLength; i++) {
     const blockReader = bufio.read(reader.readVarBytes(), true)
     const serializedBlock = readBlock(blockReader)
-    const block = node.chain.strategy.blockSerde.deserialize(serializedBlock)
+    const block = BlockSerde.deserialize(serializedBlock)
```
Comment (NullSoldier) [inline ironfish/src/rpc/routes/chain/importChain.ts:78]: I think this RPC should be deleted now, right? It can happen after this PR.

File: ironfish/src/rpc/routes/chain/importChain.test.ts [hunk 13]
```diff
@@ -19,7 +20,7 @@ describe('Route chain/importSnapshot', () => {
     Assert.isNotNull(genesis)
 
     const blockA1 = await useMinerBlockFixture(chain, 2)
-    const serializedBlockA1 = strategy.blockSerde.serialize(blockA1)
+    const serializedBlockA1 = BlockSerde.serialize(blockA1)
     const bw = bufio.write(getBlockSize(serializedBlockA1))
     const blockA1Buffer = writeBlock(bw, serializedBlockA1).render()
```
Comment (NullSoldier) [inline ironfish/src/rpc/routes/chain/importChain.test.ts:26]: I think this RPC should be deleted now, right? It can happen after this PR.

File: ironfish/src/rpc/routes/chain/snapshotChain.test.ts [hunk 14]
```diff
@@ -24,7 +25,7 @@ describe('Route chain/snapshotChainStream', () => {
       await expect(chain).toAddBlock(blockA1)
       expect(blockA1.transactions.length).toBe(1)
 
-      const serializedBlockA1 = strategy.blockSerde.serialize(blockA1)
+      const serializedBlockA1 = BlockSerde.serialize(blockA1)
       const bw2 = bufio.write(getBlockSize(serializedBlockA1))
```
Comment (NullSoldier) [inline ironfish/src/rpc/routes/chain/snapshotChain.test.ts:29]: I think this RPC should be deleted now, right? It can happen after this PR.

File: ironfish/src/utils/snapshot.ts [hunk 15]
```diff
@@ -2,9 +2,10 @@
...
  * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
 export type SnapshotManifest = {
-  block_height: number
+  block_sequence: number
   checksum: string
   file_name: string
   file_size: number
   timestamp: number
+  database_version: number
 }
```
Comment (NullSoldier) [inline ironfish/src/utils/snapshot.ts:11]: This should go in the CLI in export since it's not used by anyone other than the CLI.

File: ironfish/src/rpc/routes/chain/importChain.ts (hunk 12 above)
Comment (hughy) [reply ironfish/src/rpc/routes/chain/importChain.ts:78]: I'll delete these rpc routes and unneeded snapshot code in a subsequent PR

**Review by NullSoldier — APPROVED** [review]

**Review by NullSoldier — APPROVED** [review]
